client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -158,41 +148,38 @@ private void handleMissingLog() {
      */
     private void handleLogSealed(Segment segment) {
-        /*
-         * Using segmentSealedLock prevents concurrent segmentSealedCallback for different segments
-         * from being invoked concurrently.
-         * 
-         * By calling flush while the write lock is held we can ensure that any inflight
-         * entries that will succeed in being written to a new segment are written and any
-         * segmentSealedCallbacks that will be called happen before the next write is invoked.
-         */
-        writeFlushLock.readLock().lock();
-        Retry.indefinitelyWithExpBackoff(config.getInitalBackoffMillis(), config.getBackoffMultiple(),
-                                         config.getMaxBackoffMillis(),
-                                         t -> log.error("Encountered excemption when handeling a sealed segment: ", t))
-             .runAsync(() -> {
-                 // Running in the shrinking pool as this operation is infrequent and slow.
-                 return CompletableFuture.runAsync(() -> {
-                     segmentSealedLock.lock();
-                     log.info("Handling sealed segment {}", segment);
-                     try {
-                         resend(selector.refreshSegmentEventWritersUponSealed(segment, segmentSealedCallBack));
-                     } finally {
-                         segmentSealedLock.unlock();
-                     }
-                     /* In the case of segments merging Flush ensures there can't be anything left
-                      * inflight that will need to be resent to the new segment when the write lock
-                      * is released. (To preserve order)
+        sealedSegmentQueue.add(segment);
+        retransmitPool.execute(() -> {
+            Retry.indefinitelyWithExpBackoff(config.getInitalBackoffMillis(), config.getBackoffMultiple(),
+                                             config.getMaxBackoffMillis(),
+                                             t -> log.error("Encountered excemption when handeling a sealed segment: ", t))
+                 .run(() -> {
+                     /*
+                      * Using writeSealLock prevents concurrent segmentSealedCallback for different segments
+                      * from being invoked concurrently, or concurrently with write.
+                      * 
+                      * By calling flush while the write lock is held we can ensure that any inflight
+                      * entries that will succeed in being written to a new segment are written and any
+                      * segmentSealedCallbacks that will be called happen before the next write is invoked.
                       */
-                     flushInternal();
-                 }, retransmitPool);
-             }, backgroundThreadPool).handle((r, t) -> {
-                 writeFlushLock.readLock().unlock();
-                 return null;
-             });
+                     synchronized (writeSealLock) {
```

Comment:
My current understanding :
1. `handleLogSealed` is invoked on the netty callback thread. Hence multiple segmentsealed callbacks are mutually exclusive.
2. if there are multiple segment sealed (in event of scaledown) then multiple netty threads are blocked on `writeSealLock`. The single thread which has the lock will handle all the segment sealed requests and for each segment sealed it will flush all segments. (Note: Flush and close can be invoked concurrently here).
3. Now `SegmentOutputStream.flush()` will be blocked until that SegmentOutputstream's inflight is empty. and if this flush encounters segment sealed then the new thread would update the `sealedSegmentQueue` and wait as described in (2).

Reply:
> handleLogSealed is invoked on the netty callback thread. Hence multiple segmentsealed callbacks are mutually exclusive.

It is invoked by netty callback threads, but there is more than one such netty thread so they are not mutually exclusive. 

> Can we avoid blocking the netty threads? Earlier this operation happened on backgroundThreadPool.

It won't block them, it still does occur in a background pool. See line 152.

> if there are multiple segment sealed (in event of scaledown) then multiple netty threads are blocked on writeSealLock. The single thread which has the lock will handle all the segment sealed requests and for each segment sealed it will flush all segments. 

Yes. This is why I switched it to a background thread and a queue. That way the one thread can do them in a loop rather than having multiple threads block.

> Now SegmentOutputStream.flush() will be blocked until that SegmentOutputstream's inflight is empty. and if this flush encounters segment sealed then the new thread would update the sealedSegmentQueue and wait as described in (2)

Correct.